read_hbtplus.py

```diff
@@ -0,0 +1,354 @@
+        for ptype in sorted(list(infile)):
+            halo_size_data[ptype] = phdf5.collective_read(
+                infile[ptype]["nr_particles_bound"], comm
+            )
+
+    # Sum over particle types
+    nr_bound_part = None
+    for ptype in halo_size_data:
+        if nr_bound_part is None:
+            nr_bound_part = np.zeros_like(halo_size_data[ptype])
+        nr_bound_part += halo_size_data[ptype]
+    del halo_size_data
+
+    # Ensure that nr_bound_part is split over MPI ranks in the same way as the subhalos
+    ndesired = len(subhalo)
+    ndesired = comm.allgather(ndesired)
+    nr_bound_part = psort.repartition(nr_bound_part, ndesired, comm=comm)
+
+    # Wrap in a unyt array
+    nr_bound_part = unyt.unyt_array(
+        nr_bound_part, units=unyt.dimensionless, dtype=int, registry=registry
+    )
+
+    # Only process resolved subhalos (HBTplus also outputs unresolved "orphan" subhalos)
+    # Here we use the number of particles EXCLUDING duplicates, i.e. not nbound from HBTplus.
+    keep = nr_bound_part > 1
+
+    # Assign indexes to halos: for each halo we're going to process we store the
+    # position in the input catalogue.
+    nr_local_halos = len(keep)
+    local_offset = comm.scan(nr_local_halos) - nr_local_halos
+    index = np.arange(nr_local_halos, dtype=int) + local_offset
+    index = index[keep]
+    index = unyt.unyt_array(
+        index, units=unyt.dimensionless, dtype=int, registry=registry
+    )
+
+    # Find centre of potential
+    cofp = (
+        subhalo["ComovingMostBoundPosition"][keep, :] * LengthInMpch / h
+    ) * swift_cmpc
+
+    # Initial guess at search radius for each halo - twice the half mass radius.
+    # Search radius will be expanded if we don't find all of the bound particles.
+    search_radius = (
+        2.0 * (subhalo["RHalfComoving"][keep] * LengthInMpch / h) * swift_cmpc
+    )
+
+    # Central halo flag
+    is_central = np.where(subhalo["Rank"] == 0, 1, 0)[keep]
+    is_central = unyt.unyt_array(
+        is_central, units=unyt.dimensionless, dtype=int, registry=registry
+    )
+
+    # Subhalo tracking information
+    track_id = unyt.unyt_array(
+        subhalo["TrackId"][keep], units=unyt.dimensionless, dtype=int, registry=registry
+    )
+    host_halo_id = unyt.unyt_array(
+        subhalo["HostHaloId"][keep], units=unyt.dimensionless, dtype=int, registry=registry
+    )
+    depth = unyt.unyt_array(
+        subhalo["Depth"][keep], units=unyt.dimensionless, dtype=int, registry=registry
+    )
+    snapshot_birth = unyt.unyt_array(
+        subhalo["SnapshotIndexOfBirth"][keep], units=unyt.dimensionless, dtype=int, registry=registry
+    )
+    parent_id = unyt.unyt_array(
+        subhalo["NestedParentTrackId"][keep], units=unyt.dimensionless, dtype=int, registry=registry
+    )
+    m_bound = unyt.unyt_array(
+        subhalo["Mbound"][keep], units=unyt.dimensionless, dtype=float, registry=registry
```

Comment:
Nope good catch. I don't think we need to save this now (same for vmax and rank), since they should be the same as the SOAP values. I'll test a couple of runs to make sure this is definitely true though.